docs/src/pages/components/buttons/CustomVariantButtons.tsx

```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import {
+  createMuiTheme,
+  makeStyles,
+  Theme,
+  ThemeProvider,
+} from '@material-ui/core/styles';
+import Button from '@material-ui/core/Button';
+
+const useStyles = makeStyles((theme: Theme) => ({
+  margin: {
+    margin: theme.spacing(1),
+  },
+}));
+
+const theme = createMuiTheme({
+  variants: {
+    MuiButton: [
```

Comment:
As currently these are defined as array, if nested theme defines some other variants, those will overwrite completely the previously defined array (this is the current behavior of `deepmerge`). I added logic for merging arrays too, but that broke some other tests, so not sure what would be the best approach here. Any thoughts?

Reply:
I was thinking about adding an option `mergeArrays` which will default to false, but the problem is that under the same object - `theme` we want to handle it differently... Maybe the best option is to create a dedicated `mergeTheme` utility if we need to handle this differently..

Reply:
Yeah, looks to me that this is the safest option... @oliviertassinari what do you think of adding a dedicated `mergeTheme` utility?

Reply:
I've added `mergeTheme` functionality in the `ThemeProvider`, so variants are now correctly merged. Not sure if it is the best option, but it works in this scenario.. Please take a look

Reply:
The only alternative I could think of would be to have an object with keys. The order of the variants is important as it defines the priority. While we can also get order working with objects, it would also mean having to name each variant. Could it duplicate with the matcher?